### DISCUSSION TITLE
how does a request handler inside a controller class gets associated with the app instance?

### DISCUSSION THREAD

**Discussion by amirongit** [Neoteroi/BlackSheep#416] [Q&A] [answered]
```py
from blacksheep import Application
from blacksheep.server.controllers import Controller, get


app = Application()


class Foo(Controller):

    @get('/')
    def bar():
        return 'whatever'

```

can you see what confuses me here? how the bar method in Foo class is considered a request handler for app? (I meant how is it technically possible)

what if I had two Application instances? is it required to only have one?

**Comment by amirongit** [comment, marked as answer]
there is a singleton router which all the controllers use to register their routes. (in case anybody came across this question)

**Reply by RobertoPrevato** [reply]
Hi @amirongit 
Sorry for replying late here. Yes, a singleton router is used for controllers to reduce code verbosity. It is possible to have multiple instances of applications if desired, each using its own router. But in most use cases web services don´t use multiple instances of applications, hence that approach makes sense in my opinion.
The same principle is applied to regular routes in `v2` and now modules under `routes` and `controllers` folders are also automatically imported. The code API is much cleaner this way.

**Reply by RobertoPrevato** [reply]
And by the way, it is not correct to say that all controllers use the singleton router for their routes. BlackSheep supports multiple applications using different sets of controllers. Quite the opposite: the router for controllers is associated to an application (by default it is the singleton instantiated in the controllers module, but can be set as desired), and when one of its methods are used to decorate a method of a controller, than the controller becomes bound to the application that uses it - upon application start. This is also necessary because it is not possible to know before hand if a route registered for a controller conflicts with another route.
